### ISSUE TITLE
ER Diagram attributes

### ISSUE THREAD

**Issue by zjopy** [mermaid-js/mermaid#1361]
Thanks a ton to @spopida for adding ER diagrams to mermaid in #1308!

It would be awesome if we could add attributes of entities. As mentioned in the docs, attributes are not yet supported.

> ER diagrams are a new feature in Mermaid and are experimental. There are likely to be a few bugs and constraints, and enhancements will be made in due course. Currently you can only define entities and relationships, **but not attributes**.

Class diagrams unfortunately don't allow to specify any relations between their fields, and if I understand the distinction correctly, this would be something that ER diagrams would allow to do.

Attributes for entities would allow to make diagrams of object models like this

![er diagram](https://d2slcw3kip6qmk.cloudfront.net/marketing/pages/chart/examples/databaseerdiagram.svg)

**Comment by spopida** [comment]
Hi @zjopy - glad you like the ER diagram feature, and thanks for posting your suggestion.  Ability to add attributes is definitely something I would like to do, although I'm not quite sure when!  It's my top priority for the next enhancement (if someone doesn't beat me to it!).

Getting the lines to connect to the exact 'correct' place is likely to be challenging, because layout of the diagram is delegated to a rendering engine (dagre), so there is not so much control over where the edges (lines) intersect with the nodes (boxes).  But we'll see.

In fact I have a general concern about including foreign keys as attributes, although I know they will be popular and 'everyone does it'.  My concern is that the inclusion of a foreign key attribute turns a _logical_ ER model, that could be implemented in various ways, into a _physical_ ER model that follows the relational approach.  For example, you can implement a one-to-many relationship in JSON without the need for a foreign key using an array property embedded in the parent entity.

Of course it's not for me to stipulate why people should or shouldn't use physical or logical models, so I think it's perfectly OK to allow foreign key attributes in the diagram.  I just want to be careful not to imply that all relationships 'need' foreign key attributes.

Anyway, it's on my radar, and thanks for raising it.

**Comment by david-morris** [comment]
It looks like this is currently working, but for some reason there's no documented way to note the multiplicity of an attribute?
https://mermaid-js.github.io/mermaid/#/entityRelationshipDiagram
At least for the use case of documenting a GraphQL API, being able to write multiplicity is very important.

**Comment by jgreywolf** [comment]
@spopida Are you planning on continuing to make any enhancements to ER diagram?

**Comment by ufukty** [comment]
> Getting the lines to connect to the exact 'correct' place is likely to be challenging, because layout of the diagram is delegated to a rendering engine (dagre), so there is not so much control over where the edges (lines) intersect with the nodes (boxes). But we'll see.

I'm not sure if this is relevant: another diagraming tool is providing the functionality of drawing relations between exact columns when [ELK](https://eclipse.dev/elk/reference.html) is used as the engine. So, seems like there is another engine providing the base functionality.

**Comment by xaliander** [comment]
@spopida: I would also love this feature. The Entity Relationship Diagrams are already quite nice, but they lack the ability to directly connect rows of the ER diagrams with lines. It would be great if we could fully document our code in GitHub / GitLab including the database FKs - all in mermaid.js.

For anyone else running into this looking for a solution: 
The phpMyAdmin designer could work, alternatively "quickdatabasediagrams.com" (not affiliated) could also be an alternative.

**Comment by minev-dev** [comment]
I would also be happy to have this feature. It seems like a critical part to build good ER diagrams which should boost the Mermaid adoption.

**Comment by hannesvogel** [comment]
I think it is implemented - the documentation shows attributes the relations: https://mermaid.ai/open-source/syntax/entityRelationshipDiagram.html

**Comment by pbrolin47** [comment]
@hannesvogel You are correct. Attributes are supported in ER diagrams. In the Sample Diagram in the Live Editor, attributes are present in the diagram. Closing

**Comment by ufukty** [comment]
All of the issue body and follow-up are making me think that this issue is more about the annotated links between rows in tables rather than tables in general; which neither the docs nor the live editor currently give an example of that. 

It would probably look like this:

```
erDiagram
    CUSTOMER.ID ||--o{ ORDER.CUSTOMER_ID : places
```

**Comment by ianad** [comment]
> All of the issue body and follow-up are making me think that this issue is more about the annotated links between rows in tables rather than tables in general; which neither the docs nor the live editor currently give an example of that.
> 
> It would probably look like this:
> 
> ```
> erDiagram
>     CUSTOMER.ID ||--o{ ORDER.CUSTOMER_ID : places
> ```

@ufukty I agree. Erroneously closed imo @pbrolin47

**Comment by xaliander** [comment]
I agree with @ufukty and @ianad, I do not think this is fully resolved yet. Though the title might need updating.

What @vwkd originally showed, and what @jxu also illustrated with their images, is not just "attributes inside entities". The missing feature is a physical ER model view where specific rows/attributes can be connected directly, so the diagram shows exactly which foreign key references which key. I was also hoping for this capability.

That is different from a logical ER model, where entity-level relationships may be enough (as @spopida explained).

**Comment by pbrolin47** [comment]
Hi, 
Thanks for the feedback. Re-opened the issue since it's not just about possibility to have attributes inside entities. A rename of the issue would be good to make it clearer. Is something like "ER diagrams annotated links" a reasonable title?

**Comment by xaliander** [comment]
Thanks for reopening the issue @pbrolin47.

I think "annotated links" is a bit vague. The missing feature is not just links with annotations, but the ability to connect a relationship to specific attributes/fields within entities.

I think a clearer title could be:
"ER diagrams: connecting specified rows (e.g. FK to PK)"
or
"ER diagrams: links between specific attributes (e.g. FK to PK)"

**Comment by pbrolin47** [comment]
Thanks @xaliander. Just renamed issue
